### PR TITLE
feat: rewrite from scratch [BREAKING CHANGES]

### LINKED ISSUE #20: Replace underlying Fetch API implementation
**Is your feature request related to a problem? Please describe.**
[fetch-h2](https://github.com/grantila/fetch-h2), the current underlying Fetch API implementation, has a serious scalability bottleneck: Parallel requests to the same HTTP/2 origin are effectively serialised, thus defeating the purpose of HTTP/2 Multiplexing. See #14.

**Describe the solution you'd like**
A. Replacing `fetch-h2` with another Fetch API implementation which transparently supports HTTP/1(.1) and HTTP/2 and which provides better scalability for parallel requests. [node-fetch-h2](https://github.com/MikeRalphson/node-fetch-h2/tree/h2) seems to be a good candidate.

B. Alternatively, implement the relevant parts of the Fetch API and write a simple abstraction of the builtin `http`, `https` and `http2` modules tailored to our needs.

**Describe alternatives you've considered**
I've created a `fetch-h2` [issue](https://github.com/grantila/fetch-h2/issues/85) and looked into creating a PR. The code however is hardly documented and the flow hard to follow. The maintainer doesn't seem to be very responsive either.

### PR DESCRIPTION
**Pull request by stefan-guggisberg**
The underlying Fetch implementation is rewritten from scratch, as the the previously used implementation (fetch-h2) has serious long-standing issues and its maintainer is unresponsive and doesn't seem to care. 

BREAKING CHANGE: The exposed API, options etc have been aligned with node-fetch where possible. Due to the major
changes there are numerous breaking changes both in the API and the options.

## Related Issues

fixes #20, #27, #32, #81

## TODOs:

* [x] improve doc: add more examples, describe breaking changes/migration tips, etc (`README.md`)
* [x] improve code coverage, add more tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,52 +1,209 @@
...
 * `Response.buffer()` returns a Node.js `Buffer`.
-* The `body` that can be sent in a `Request` can also be a `Readable` Node.js stream, a `Buffer` or a string.
-* `fetch()` has an extra option `json` that can be used instead of `body` to send an object that will be JSON stringified. The appropriate content-type will be set if it isn't already.
-* ~~`fetch()` has an extra option `timeout` which is a timeout in milliseconds before the request should be aborted and the returned promise thereby rejected (with a `TimeoutError`).~~
+* `Response.url` contains the final url when following redirects.
+* The `body` that can be sent in a `Request` can also be a `Readable` Node.js stream, a `Buffer`, a string or a plain object.
+* There are no forbidden header names.
+* The `Response` object has an extra property `httpVersion` which is one of `'1.0'`, `'1.1'` or `'2.0'`, depending on what was negotiated with the server.
+* The `Response` object has an extra property `fromCache` which determines whether the response was retrieved from cache.
+* `Response.headers.plain()` returns the headers as a plain object.
```
Comment (tripodsan) [inline README.md:73]: why not `raw()` like node-fetch?  https://github.com/node-fetch/node-fetch/blob/4abbfd231f4bce7dbe65e060a6323fc6917fd6d9/src/headers.js#L200

File: src/fetch/headers.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+class Headers {
+  /**
+   * Constructs a new Headers instance
+   *
+   * @constructor
+   * @param {Object} [init={}]
+   */
+  constructor(init = {}) {
+    this[INTERNALS] = {
+      map: {},
```
Comment (tripodsan) [inline src/fetch/headers.js:73]: why not just use the `Map()` object? many functions entries/values will come for free

**Review by tripodsan — COMMENTED** [review]
as far as I can see,

`index.d.ts` exports Request, Response etc, but
`index.js` doesn't.

eg, if I want to use `Response`, I need to to:

```js
const Response = require('@adobe/helix-fetch/request/response.js');
```

maybe it would be easier to:

```js
const { Response } = require('@adobe/helix-fetch');
```

If Request, Response should not be used directly, they shouldn't be exported in `index.d.ts`.

(but it would be great, if we could use it in the unified actions, instead of node-fetch)

File: README.md (hunk 1 above)
Comment (stefan-guggisberg) [reply README.md:73]: That's exactly the reason for renaming it to `plain()` ;-) `node-fetch`'s `raw()` method returns all headers and their values as array whereas `plain()` returns the headers as a plain JS object.

**Comment by stefan-guggisberg** [comment]
> as far as I can see,
> 
> `index.d.ts` exports Request, Response etc, but
> `index.js` doesn't.
> 
> eg, if I want to use `Response`, I need to to:
> 
> ```js
> const Response = require('@adobe/helix-fetch/request/response.js');
> ```
> 
> maybe it would be easier to:
> 
> ```js
> const { Response } = require('@adobe/helix-fetch');
> ```
> 
> If Request, Response should not be used directly, they shouldn't be exported in `index.d.ts`.
> 
> (but it would be great, if we could use it in the unified actions, instead of node-fetch)

sure.  `Request`, Response`, `Headers` etc are already exported:

```js
const { Request, Response } = require('@adobe/helix-fetch');
```

File: README.md (hunk 1 above)
Comment (tripodsan) [reply README.md:73]: a, right. thanks.

**Comment by tripodsan** [comment]
> sure. `Request`, `Response`, `Headers` etc are already exported:
> 
> ```js
> const { Request, Response } = require('@adobe/helix-fetch');
> ```
ah sorry, I missed it when looking just at the diff:
https://github.com/adobe/helix-fetch/blob/3dbef15ce8c1d1ab45ad17dc216d5842bec01e02/src/fetch/index.js#L395

File: src/fetch/headers.js (hunk 2 above)
Comment (stefan-guggisberg) [reply src/fetch/headers.js]: Good point. However, while changing the implementation to use `Map` internally I remember why I did't in the first place: `Map` and its iterators are insertion ordered while the `Headers` iterators are alphabetically ordered. In the end the benefit of using `Map` versus `{}` is debatable. Never mind, here's the change: https://github.com/adobe/helix-fetch/pull/116/commits/2c13768b9fb9b7c095fd75c471d67f8d4fad5845

File: src/fetch/headers.js [hunk 3]
```diff
@@ -118,11 +118,9 @@ class Headers {
     }
   }
 
-  * keys() {
-    // return Object.keys(this[INTERNALS].map).sort();
-    for (const name of Object.keys(this[INTERNALS].map).sort()) {
-      yield name;
-    }
+  keys() {
```
Comment (tripodsan) [inline src/fetch/headers.js:121]: why not just `return this[INTERNALS].map.keys()` ?

File: src/fetch/headers.js [hunk 4]
```diff
@@ -118,11 +118,9 @@ class Headers {
...
+    return Array.from(this[INTERNALS].map.keys())
+      .sort();
   }
 
   * values() {
```
Comment (tripodsan) [inline src/fetch/headers.js:126]: why not just `return this[INTERNALS].map.values()` ?

File: src/fetch/headers.js [hunk 5]
```diff
@@ -158,7 +156,15 @@ class Headers {
    * @return {object}
    */
   plain() {
-    return { ...this[INTERNALS].map };
```
Comment (tripodsan) [inline src/fetch/headers.js]: +1 this is anyways better, so you can't just modify the pojo :-)

File: src/fetch/headers.js [hunk 6]
```diff
@@ -158,7 +156,15 @@ class Headers {
...
+    if (typeof Object.fromEntries === 'function') {
+      // since node 12
+      return Object.fromEntries(this[INTERNALS].map);
```
Comment (tripodsan) [inline src/fetch/headers.js:162]: oh, I didn't know.... so useful !

File: src/fetch/headers.js (hunk 3 above)
Comment (stefan-guggisberg) [reply src/fetch/headers.js:121]: `Map` is insertion ordered while the iterator needs to return the keys in alphabetical order.

File: src/fetch/headers.js (hunk 4 above)
Comment (stefan-guggisberg) [reply src/fetch/headers.js:126]: Same here: because `Map` is insertion ordered and the values are expected ordered by keys. Have a look at the `node-fetch` test cases or try it out in the browser.

File: src/fetch/headers.js (hunk 3 above)
Comment (tripodsan) [reply src/fetch/headers.js:121]: > `Map` is insertion ordered while the iterator needs to return the keys in alphabetical order.

why ?

File: src/fetch/headers.js (hunk 4 above)
Comment (tripodsan) [reply src/fetch/headers.js:126]: ah, I don't know the spec says they have to be sorted.

Comment (stefan-guggisberg) [reply src/fetch/headers.js:126]: the spec isn't crystal clear in this regard but `node-fetch`, Chrome and Safari do return the headers in lexicographical order.
